modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java

```diff
@@ -1225,6 +1232,82 @@ private void distributedExchange() throws IgniteCheckedException {
         initDone();
     }
 
+    /**
+     *
+     */
+    private void onDistributedExchangeTimeout() {
+        if (!isDone()) {
+            synchronized (mux) {
+                if (crd.isLocal()) {
+                    Set<UUID> remaining0 = new HashSet<>(this.remaining);
+
+                    log.warning("Exchange timeout reached on coordinator node, kicking pending nodes: " +
+                        S.arrayToString(UUID.class, remaining0.toArray()));
+
+                    for (UUID remainingUuid : remaining0)
```

Comment:
ticket in jira says: " it should check that stopping these nodes won't lead to data loss and then stop them forcibly."
I don't see how you check this here.